File: Jena/operation/graph/PageRank.py
'''
Created on Jan 21, 2016

@author: phantom
'''
from operation.GraphOperation import GraphOperation

class PageRank(GraphOperation):
    COMMAND_TO_INITIALIZE_WORKING_GRAPH = '''
            CREATE GRAPH <http://workingGraph>;
                                          '''
    COMMAND_TO_FINALIZE_WORKING_GRAPH = '''
            DROP GRAPH <http://workingGraph>;
                                        '''
    COMMAND_TO_COUNT_NODES = '''
            SELECT (COUNT(DISTINCT ?node) AS ?count)
            WHERE {
                ?node <urn:connectedTo>|^<urn:connectedTo> ?neighbor
            }
                             '''
    COMMAND_TO_RETRIEVE_OUTDEGREE = '''
            INSERT { GRAPH <http://workingGraph> { ?node <temp:outDegree> ?outDegree }}
            WHERE {
                {
                    SELECT ?node (COUNT(*) AS ?outDegree)
                    WHERE {
                        ?node <urn:connectedTo> ?neighbor
                    }
                    GROUP BY ?node
                }
                UNION
                {
                    SELECT ?node ?outDegree
                    WHERE {
                        ?neighbor <urn:connectedTo> ?node .
                        OPTIONAL { ?node <urn:connectedTo> ?any } .
                        FILTER (!BOUND(?any)) .
                        BIND (0 AS ?outDegree)
                    }
                }
            };
                                    '''
    COMMAND_TO_SET_INITIAL_PAGERANKS = '''
            INSERT {{ GRAPH <http://workingGraph> {{ ?node <temp:rank> {RANK} }}}}
            WHERE {{
                GRAPH <http://workingGraph> {{
                    ?node <temp:outDegree> []
                }}
            }};
                                       '''
    COMMAND_TO_UPDATE_PAGERANKS = '''
            DELETE {{ GRAPH <http://workingGraph> {{ ?node <temp:rank> ?previousRank }}}}
            INSERT {{ GRAPH <http://workingGraph> {{ ?node <temp:rank> ?newRank; <temp:delta> ?delta }}}}
            WHERE {{
                {{
                    SELECT ?node (SUM(?score) * {DAMPING_FACTOR} + (1.0 - {DAMPING_FACTOR}) / {NODES} AS ?newRank)
                    WHERE {{
                        GRAPH <http://workingGraph> {{
                            ?neighbor <temp:rank> ?rank; <temp:outDegree> ?outDegree .
                            BIND ((?rank / ?outDegree) AS ?score)
                        }} .
                        ?neighbor <urn:connectedTo> ?node .
                    }}
                    GROUP BY ?node
                }} .
                GRAPH <http://workingGraph> {{
                    ?node <temp:rank> ?previousRank
                }} .
                BIND (ABS(?previousRank - ?newRank) AS ?delta)
            }};
                                  '''
    COMMAND_TO_CHECK_CONVERGENCE = '''
            SELECT (SUM(?delta) AS ?epsilon)
            WHERE {
                GRAPH <http://workingGraph> {
                    ?node <temp:delta> ?delta
                }
            }
                                   '''
    COMMAND_TO_DELETE_DELTAS = '''
            DELETE { GRAPH <http://workingGraph> { ?node <temp:delta> ?delta }}
            WHERE {
                GRAPH <http://workingGraph> {
                    ?node <temp:delta> ?delta
                }
            };
                               '''
    COMMAND_TO_RETRIEVE_PAGERANKS = '''
            SELECT ?node ?rank
            WHERE {
                GRAPH <http://workingGraph> {
                    ?node <temp:rank> ?rank
                }
            }
                                    '''

    def __init__(self, epsilon, dampingFactor, connection):
        GraphOperation.__init__(self, 'PageRank', connection)
        self.epsilon = epsilon
        self.dampingFactor = dampingFactor
    
    def initialize(self):
        self.connection.urika.update(self.name, PageRank.COMMAND_TO_INITIALIZE_WORKING_GRAPH)
        self.connection.urika.update(self.name, PageRank.COMMAND_TO_RETRIEVE_OUTDEGREE)
        results = self.connection.urika.query(self.name, PageRank.COMMAND_TO_COUNT_NODES, None, None, 'json', True)
        for result in results['results']['bindings']:
            self.numberOfNodes = int(result['count']['value'])
        
        self.connection.urika.update(self.name, PageRank.COMMAND_TO_SET_INITIAL_PAGERANKS.format(RANK = (1.0 / self.numberOfNodes)))
    
    def process(self):
        self.iteration = 0
        
        while True:
            self.iteration += 1
            print '\tIteration: ' + str(self.iteration)
            
            self.connection.urika.update(self.name, PageRank.COMMAND_TO_UPDATE_PAGERANKS.format(DAMPING_FACTOR = self.dampingFactor, NODES = self.numberOfNodes))
            
            if self.converges():
                break
            
    def store(self):
        print '\tNumber of Iteration: ' + str(self.iteration)
        
        results = self.connection.urika.query(self.name, PageRank.COMMAND_TO_RETRIEVE_PAGERANKS)
        for result in results['results']['bindings']:
            self.writer.writeLine('Node: ' + result['node']['value'] + " [PageRank: " + result['rank']['value'] + "]")
    
    def finalize(self):
        self.connection.urika.update(self.name, PageRank.COMMAND_TO_FINALIZE_WORKING_GRAPH)
    
    def converges(self):
        results = self.connection.urika.query(self.name, PageRank.COMMAND_TO_CHECK_CONVERGENCE)
        for result in results['results']['bindings']:
            epsilon = float(result['epsilon']['value'])
        
        self.connection.urika.update(self.name, PageRank.COMMAND_TO_DELETE_DELTAS)
        
        return epsilon < self.epsilon

File: Sesame/src/operation/graph/PageRank.java
package operation.graph;

import org.openrdf.model.Literal;
import org.openrdf.query.BindingSet;
import org.openrdf.query.TupleQueryResult;
import org.openrdf.repository.RepositoryConnection;

import exception.OperationException;
import operation.GraphOperation;

public class PageRank extends GraphOperation {
	private static final String NAME = "PageRank";
	
	private static final String QUERY_INIT_GRAPH = ""
			+ "CREATE GRAPH <http://workingGraph>;";
	private static final String QUERY_NODE_COUNT = ""
			+ "SELECT (COUNT(DISTINCT ?node) AS ?count)\n"
			+ "WHERE {\n"
			+ "  ?node <urn:connectedTo>|^<urn:connectedTo> ?neighbor\n"
			+ "}";
	private static final String QUERY_OUTDEGREES = ""
			+ "INSERT { GRAPH <http://workingGraph> { ?node <temp:outDegree> ?outDegree }}\n"
			+ "WHERE {\n"
			+ "  {\n"
			+ "    SELECT ?node (COUNT(*) AS ?outDegree)\n"
			+ "    WHERE {\n"
			+ "      ?node <urn:connectedTo> ?neighbor\n"
			+ "    }\n"
			+ "    GROUP BY ?node\n"
			+ "  }\n"
			+ "  UNION\n"
			+ "  {\n"
			+ "    SELECT ?node ?outDegree\n"
			+ "    WHERE {\n"
			+ "      ?neighbor <urn:connectedTo> ?node .\n"
			+ "      OPTIONAL { ?node <urn:connectedTo> ?any } .\n"
			+ "      FILTER (!BOUND(?any)) .\n"
			+ "      BIND (0 AS ?outDegree)\n"
			+ "    }\n"
			+ "  }\n"
			+ "};";
	private static final String QUERY_INITIAL_PAGERANKS = ""
			+ "INSERT { GRAPH <http://workingGraph> { ?node <temp:rank> %f }}\n"
			+ "WHERE {\n"
			+ "  GRAPH <http://workingGraph> {\n"
			+ "    ?node <temp:outDegree> []\n"
			+ "  }\n"
			+ "};";
	private static final String QUERY_UPDATE_PAGERANKS = ""
			+ "DELETE { GRAPH <http://workingGraph> { ?node <temp:rank> ?previousRank }}\n"
			+ "INSERT { GRAPH <http://workingGraph> { ?node <temp:rank> ?newRank; <temp:delta> ?delta }}\n"
			+ "WHERE {\n"
			+ "  {\n"
			+ "    SELECT ?node (SUM(?score) * %f + (1.0 - %f) / %d AS ?newRank)\n"
			+ "    WHERE {\n"
			+ "      ?neighbor <urn:connectedTo> ?node .\n"
			+ "      GRAPH <http://workingGraph> {\n"
			+ "        ?neighbor <temp:rank> ?rank; <temp:outDegree> ?outDegree .\n"
			+ "        BIND ((?rank / ?outDegree) AS ?score)\n"
			+ "      }\n"
			+ "    }\n"
			+ "    GROUP BY ?node\n"
			+ "  } .\n"
			+ "  GRAPH <http://workingGraph> {\n"
			+ "    ?node <temp:rank> ?previousRank\n"
			+ "  } .\n"
			+ "  BIND (ABS(?previousRank - ?newRank) AS ?delta)\n"
			+ "};";
	private static final String QUERY_EPSILON = ""
			+ "SELECT (SUM(?delta) AS ?epsilon)\n"
			+ "WHERE {\n"
			+ "  GRAPH <http://workingGraph> {\n"
			+ "    ?node <temp:delta> ?delta\n"
			+ "  }\n"
			+ "}";
	private static final String QUERY_DELETE_EPSILON = ""
			+ "DELETE { GRAPH <http://workingGraph> { ?node <temp:delta> ?delta }}\n"
			+ "WHERE {\n"
			+ "  GRAPH <http://workingGraph> {\n"
			+ "    ?node <temp:delta> ?delta\n"
			+ "  }\n"
			+ "};";
	private static final String QUERY_RETRIEVE_PAGERANKS = ""
			+ "SELECT ?node ?rank\n"
			+ "WHERE {\n"
			+ "  GRAPH <http://workingGraph> {\n"
			+ "    ?node <temp:rank> ?rank\n"
			+ "  }\n"
			+ "}";
	private static final String QUERY_DROP_GRAPH = ""
			+ "DROP GRAPH <http://workingGraph>;";
	
	private int numberOfNodes;
	private float dampingFactor;
	private float epsilon;
	
	public PageRank(RepositoryConnection connection, float epsilon, float dampingFactor) throws OperationException {
		super(NAME, connection);
		this.dampingFactor = dampingFactor;
		this.epsilon = epsilon;
	}
	
	@Override
	protected void init() {
		numberOfNodes = computeNumberOfNodes();
		connection.prepareUpdate(QUERY_INIT_GRAPH).execute();
		connection.prepareUpdate(QUERY_OUTDEGREES).execute();
		connection.prepareUpdate(String.format(QUERY_INITIAL_PAGERANKS, (1.0f / numberOfNodes))).execute();
	}

	@Override
	protected void process() {
		int iteration = 0;
		
		do {
			iteration++;
			System.out.println("\tIteration: " + iteration);
			connection.prepareUpdate(String.format(QUERY_UPDATE_PAGERANKS, this.dampingFactor, this.dampingFactor, this.numberOfNodes)).execute();
		} while (!converges());
		
		System.out.println("The number of iterations: " + iteration);
		savePageRanks();
	}

	@Override
	protected void end() {
		connection.prepareUpdate(QUERY_DROP_GRAPH).execute();
	}

	private int computeNumberOfNodes() {
		int count = 0;
		TupleQueryResult result = connection.prepareTupleQuery(QUERY_NODE_COUNT).evaluate();
		if(result.hasNext()) {
			BindingSet binding = result.next();
			count = ((Literal)binding.getValue("count")).intValue();
		}
		return count;
	}
	
	private boolean converges() {
		float epsilon = 0.0f;
		TupleQueryResult result = connection.prepareTupleQuery(QUERY_EPSILON).evaluate();
		if(result.hasNext()) {
			BindingSet binding = result.next();
			epsilon = ((Literal)binding.getValue("epsilon")).floatValue();
		}
		System.out.println("\tEpsilon: " + epsilon);
		connection.prepareUpdate(QUERY_DELETE_EPSILON).execute();
		return epsilon < this.epsilon;
	}
	
	private void savePageRanks() {
		TupleQueryResult result = connection.prepareTupleQuery(QUERY_RETRIEVE_PAGERANKS).evaluate();
		while(result.hasNext()) {
			BindingSet binding = result.next();
			String node = binding.getValue("node").toString();
			float rank = ((Literal)binding.getValue("rank")).floatValue();
			
			writer.writeLine("Node: " + node + " [PageRank: " + (String.format("%.7f", rank)) + "]");
		}
	}
}
